### DISCUSSION TITLE
Frontend POAP Web App

### DISCUSSION THREAD

**Discussion by jonathanlei** [XRPLBounties/Proposals#119] [Ideas]
Develop a user-friendly no-code solution web app that allows event organizers to distribute attendance NFTs to participants at XRPL events, leveraging the previously developed [Proof of Attendance Infrastructure](https://github.com/XRPLBounties/Proof-of-Attendance) for the XRP Ledger. The web app will specifically utilize the API library provided by the Proof of Attendance Infrastructure for minting and distributing NFTs, as well as handling other related functionalities.

**Comment by JustAnotherDevv** [comment]
@jonathanlei I have a few questions about the bounty. The pr was merged and the md file detailing bounty is already in the `open_bounties` but link to that file isn't the readme. With all that in mind is it okay to start working on the solution already?

Also the previous bounty for the Proof Of Attendance still has last milestone that requires development of PoA web app. Is that milestone no longer active because of this or would correct solution in theory be considered for both bounties?

**Comment by jonathanlei** [comment]
Hello @JustAnotherDevv
The frontend POAP bounty is ready for submission, feel free to start working on it! 
and thanks for spotting the difference, I created a PR to update the README and closing the backend POAP bounty, should be merged soon.

**Comment by jonathanlei** [comment]
Please submit by creating a PR on https://github.com/XRPLBounties/POAP-APP, thanks!

**Comment by TusharPardhe** [comment]
@jonathanlei For the bounty I think UI/UX Design should be the first in the line as after the finalisation of design one can start coding/integrating the APIs. Also, I would like to ask that for designs is there a specific guideline of using something like for eg: Figma or it can be just a raw outline, or we can implement on the fly and work on the design based on feedback?

**Reply by jonathanlei** [reply]
Hey! 
A raw outline would be fine, and happy to provide feedback along the way

**Comment by JustAnotherDevv** [comment]
Hi, I have created my own submission for the bounty [here](https://github.com/XRPLBounties/POAP-APP/pull/3).

**Comment by JST5000** [comment]
Hello @TusharPardhe, @rikublock, and @JustAnotherDevv,

Firstly, we want to express our gratitude for the effort you've all invested into this bounty project so far. Your commitment is much appreciated.

Given the nature of this project and the fact that only one submission can be chosen, we've decided to make this selection now to minimize duplicated work. **After careful evaluation of all PRs, we've decided to proceed with @rikublock's submission, provided they're comfortable with completing milestones 4 & 5.**

From our analysis of your submissions, it's become clear that milestones 4 & 5 are necessary for end users to use the tool. Without these, attendees could potentially mint NFTs as if they were event organizers, locking all the XRP in the minting account. Moreover, attendees will end up seeing many complicated features that they don’t need. To address this, it's crucial to distinguish the login process for those claiming an event NFT from those creating an event.

Another challenge we saw was that as written, it’s unclear how non-technical event organizers can use the app to mint. Currently it seems that in order to provide the minting account’s secret the event organizer has to be running the backend themselves. That would require a decent amount of technical knowledge to set up. We’d love help figuring out the best way to do this.

Some ideas that we considered:
1. Hosting this service and having event organizers provide a wallet for minting, then minting NFTs from the front-end instead of the backend so the secret is never transmitted over the internet. (The downside of this is that we may be encouraging a user to share a secret to a 3rd party app which is bad practice)
2. Having the event organizer send XRP to the minting account for the hosted service to cover the XRP reserve requirements, then having the service return the XRP as NFTokens are given out. (And potentially after a certain time period burning all NFTs that weren’t transferred and returning the remaining reserve XRP)
3. It might work to issue NFTs on the event organizer’s behalf - https://xrpl.org/nftokenmint.html#issuing-on-behalf-of-another-account 

The first two ideas come with serious trade-offs, so the 3rd option seems the most promising. Open to thoughts on the best way to go about this so that non-technical event organizers can reasonably use this tool at events going forward. 

Finally, we apologize for the ambiguity around the necessity of completing milestones 4 & 5 for claiming this bounty. Their completion is essential to the usability of this tool which is ultimately what this bounty is about, and we thank you for your understanding.

**Reply by rikublock** [reply]
> After careful evaluation of all PRs, we've decided to proceed with @rikublock's submission, provided they're comfortable with completing milestones 4 & 5.

I'm open to completing these milestones, but would need more detailed requirements. It's just not clear enough to me from [these](https://github.com/XRPLBounties/Proposals/blob/main/open_bounties/0119%20Frontend%20POAP%20Web%20APP.md) bullet points alone (things like user flow, login process, claim process, etc.)

> The first two ideas come with serious trade-offs, so the 3rd option seems the most promising. Open to thoughts on the best way to go about this so that non-technical event organizers can reasonably use this tool at events going forward.

The problem I see with the 3rd option, how would we distribute the NFTs ? Even if we could create NFT sell order transactions on behalf of the organizer, they would still have to confirm them. For a 100 slot event that would mean reviewing and confirming 100 transaction requests in a Xumm wallet. 

The most practical approach I can see: The event management platform (backend and frontend) is hosted and run by a third party. Both organizers and attendees are just users on the platform. This makes it easily accessible for anyone, including non-technical event organizers. The platform takes care of ledger owner reserves. All operations are performed from a single vault wallet account operated and owned by the platform. In that case it would come down to managing reserves.

**Breakdown of reserve requirements, example 100 slots event:**
- 8-10 XRP for NFT object reserves, assuming full pre-mint (peak, decreases as users claim them)
- estimate 10-20 XRP for sell offer reserves (5-10% of total slots)
  - create sell offers on demand when user requests NFT
  - but don't cancel it, if a user fails to immediately claim the offer
- burn remaining NFTs / delete open offers when an event is expired/closed/finished (frees up any dangling reserves)
- the platform has to match reserves depending on active organizers

In short: Would need about 1/4 of slots as reserves in XRP. 

**Avoid abuse:**
- limit the maximum number of slots for a single event (e.g. 200)
- limit the total number of slots an organizer account can use (e.g. 200)
  - an organizer can create 2 events with 90 slots each, but adding a 3rd one with 30 slots would fail (90 + 90 + 30 > 200)
  - creating 10 events with 15 slots each would work (total 150 < 200)
  - slots would free up again once an event is closed or canceled
  - that number could be raised for certain users depending on requirements

**Reply by JST5000** [reply]
Managing reserves that way makes sense to me. 

Regarding reserves for sell offers, we could potentially limit the max # of concurrent sell offers. If more than that amount are opened, potentially close the oldest outstanding sell offer if it's older than X minutes, otherwise reject the incoming claim telling them to try again later. That way you could guarantee you won't run into your max reserve limit, and you can also avoid getting spammed with requests to exhaust reserves. 

I'll aim to get back to you this week with more detailed requirements for milestones 4 & 5. I'll need a bit of time to think through what would be needed there :) (Open to input if you've got some thoughts on what that could look like)

**Reply by rikublock** [reply]
> Open to input if you've got some thoughts on what that could look like

I like to use the wallet (address) as the user identifier and integrate it in the login process (like a web3 app). Makes things simple for a user. Then display UI elements/features depending on user role and access level. 

The latest version of my Dashboard App PR follows that approach. Did you already have a chance to test it ?

**Reply by JST5000** [reply]
Haven't had a chance for the most up to date version yet. Was planning on reviewing it tomorrow (Thursday).

**Reply by VEER-TANMAY-SACHIN** [reply]
is this closed ?

**Reply by JST5000** [reply]
XRPL Bounties is no longer accepting any new submissions as the program is sunset. There's still some remaining work on this POAP bounty that @rikublock has been doing, and we've been working to close this out as the last bounty :) 

(If you're curious, you can see their work here: https://github.com/XRPLBounties/POAP-APP/pull/2 )
